doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+# Events reported from SONiC as structured
+
+## Goals
+1. Provide a unified way for listing identified alertable events in SONiC switch.
+2. Provide a unified way for event detectors to report the events.
+3. Enforce a structured format for event data with pre-defined schema.
+4. Have the ability to stream at the max of 10K events per second to external clients.
+5. Ensure events are immutable across SONiC releases, but can be deprecated.
+6. Meet the reliability goal of 99.5% - From event generated to end client.
+7. Rate of event reporting can be in par with rate of syslog reporting.
+
+
+## Problems to solve
+The external tools that monitor system health often use syslog messages to look for events that need alert raised.
+The syslog messages being text string could potentially **change** across releases. Some log messages could get split into multiple different ones.
+This poses a challenge to the external tools as they are forced to adapt for multiple different versions and parse a message based on the OS version of the producer.
+This would become even more challenging, when we upgrade individual container images via kubernetes/app-package-manager.
+
+The tools that use syslog messages face higher latency, as they have to wait for syslog messages to arrive to an external repository. 
+This latency could run in the order of minutes.
+
+
+![image](https://user-images.githubusercontent.com/47282725/156947460-66d08b3d-c981-4413-b0d5-232643dfba01.png)
+
+
+![image](https://user-images.githubusercontent.com/47282725/159061158-30ff3c8a-5fc0-4af2-8822-6bfc67b2329c.png)
+
+
+
+## Requirements
+### Events
+1. Events are defined with schema.
+2. Every event is identified by tag, which is unique within a process with zero or more event specific parameters.
+3. Events are static (*don't change*) across releases, but can be deprecated in newer releases.
+4. Every event is described in YANG schema.
+5. YANG schema files for all events are available in a single location for NB clients to refer in Switch.
+
+### Event detection
+1. The method of detection can be any.
+2. This can vary across events.
+3. Syslog messages might be used to detect an event or some custom queries or code-update to report events directly or ...
```

Comment:
>code-update to report events directly [](http://example.com/codeflow?start=78&length=37)

Could you explain "code-update to report events directly"?

Reply:
The code we own, like orchagent, syncd, all the monit scripts and similar, could call the API/macro directly to report the event.